### ISSUE TITLE
Transfer a large amount of data will crash the docker daemon

### ISSUE THREAD

**Issue by jack230230** [docker/docker#14460]
docker version:
```
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64
```

docker info:
```
Containers: 518
Images: 670
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1706
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-53-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 2
Total Memory: 3.861 GiB
Name: vagrant-ubuntu-trusty-64
ID: N6ET:QS6W:OP6J:5ANI:7Q3B:E7NM:QBAP:FAVH:3YTX:LYMR:XJSL:EZPY
Debug mode (server): true
File Descriptors: 11
Goroutines: 15
System Time: 2015-07-08T17:43:59.432060733+08:00
EventsListeners: 0
Init SHA1:
Init Path: /usr/bin/docker
Docker Root Dir: /var/lib/docker
WARNING: No swap limit support
```
uname -a:

```
Linux vagrant-ubuntu-trusty-64 3.13.0-53-generic #89-Ubuntu SMP Wed May 20 10:34:39 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

When I attach the container stdout and transfer a large amount of data, docker daemon crashed ,and I get error log of docker daemon.
```
INFO[0378] POST /v1.19/containers/create
ERRO[0378] Couldn't run auplink before unmount: fork/exec /sbin/auplink: cannot allocate memory
ERRO[0378] Couldn't run auplink before unmount: fork/exec /sbin/auplink: cannot allocate memory
DEBU[0378] Calling POST /containers/{name:.*}/attach
INFO[0378] POST /v1.19/containers/743fd2471a0be51700b0fbc0e5cea53cb2840da5b0e91d12452743f052840ac4/attach?stderr=1&stdin=1&stdout=1&stream=1
DEBU[0378] attach: stdin: begin
DEBU[0378] attach: stdout: begin
DEBU[0378] attach: stderr: begin
DEBU[0378] Calling POST /containers/{name:.*}/start
INFO[0378] POST /v1.19/containers/743fd2471a0be51700b0fbc0e5cea53cb2840da5b0e91d12452743f052840ac4/start
ERRO[0378] leaving endpoint failed: a container has already joined the endpoint
ERRO[0378] Couldn't run auplink before unmount: fork/exec /sbin/auplink: cannot allocate memory
ERRO[0378] Handler for POST /containers/{name:.*}/start returned error: Cannot start container 743fd2471a0be51700b0fbc0e5cea53cb2840da5b0e91d12452743f052840ac4: namespace creation reexec command failed: fork/exec /usr/bin/docker: cannot allocate memory
ERRO[0378] HTTP Error                                    err=Cannot start container 743fd2471a0be51700b0fbc0e5cea53cb2840da5b0e91d12452743f052840ac4: namespace creation reexec command failed: fork/exec /usr/bin/docker: cannot allocate memory statusCode=500
DEBU[0378] Closing buffered stdin pipe
DEBU[0378] attach: stdin: end
```
You can run these test cmds:

```
docker run -i --log-driver=none ubuntu dd if=/dev/zero of=/proc/self/fd/1 bs=1M count=1000
```

```
docker run -i --log-driver=none ubuntu yes hellwflwejfwefklefjkejfkejfkejfigjrijgrijgijfejfejfjejfefjefjeijfiejfiejfiejfijeifjeifjeijfiejfiej >> hello.txt
```

**Comment by thefallentree** [comment]
We suspect docker daemon is leaking memory somewhere copying stdout to hijack connection.

**Comment by hqhq** [comment]
According to your error message, you just run out of memory.

**Comment by jack230230** [comment]
@hqhq I have 4G memory (3G for free), but dd just generate 1G data ,why run out of memory ?

**Comment by thefallentree** [comment]
@hqhq 

The problem we are describing is that  there seems to be memory bloating issue in docker daemon when executing commands that outputs huge amounts of data, (Even without enabling logging)  This is a issue for us because we are trying to replace a command SSH with docker exec.

**Comment by hqhq** [comment]
Yes, I see the problem, I'm investigating now.

**Comment by coolljt0725** [comment]
Just see this issue, the reason is that https://github.com/docker/docker/blob/master/pkg/ioutils/readers.go#L76 the `buf` is a variable-sized buffer, so https://github.com/docker/docker/blob/master/pkg/ioutils/readers.go#L187 will keep on writing to this `buf` until the memory run out.

**Comment by thaJeztah** [comment]
including @unclejack in this conversation as he's currently working on some memory related issues

ping @unclejack ^^

**Comment by coolljt0725** [comment]
In normal the buf will not keep on growing because some process will read from the buf and clear the buf, but at this scenario `docker run -i --log-driver=none ubuntu yes hellwflwejfwefklefjkejfkejfkejfigjrijgrijgijfejfejfjejfefjefjeijfiejfiejfiejfijeifjeifjeijfiejfiej >> hello.txt` the data will write to `buf` first and then write to the disk, but the speed write to disk is much slower than the container write to the `buf`, so the `buf` keep on growing until the memory run out.

**Comment by unclejack** [comment]
@thaJeztah Thanks for the ping!

@coolljt0725 Yes, this is a known problem right now. The fix will be quite complicated and it needs to be done after the other major problems with memory usage are fixed. I've got to try out some ideas and propose a fix which doesn't impact performance in any way.

That unbounded and endless buffer growth has been fixed for non-degenerate cases (long running containers which write 0.1-100 KB/s to stdout/stderr).

This scenario with a flood is the last issue which needs to be fixed as far as I know.

**Comment by thefallentree** [comment]
An interesting observation is that if we enable pty, docker will not crash .

```docker run -it --log-driver=none ubuntu yes hellwflwejfwefklefjkejfkejfkejfigjrijgrijgijfejfejfjejfefjefjeijfiejfiejfiejfijeifjeifjeijfiejfiej >> hello.txt```

My guess is that PTY have some sort of flow control, is that the case ?

**Comment by thefallentree** [comment]
@unclejack  Another thing is that the memory is not immediately released upon ```docker run -i --log-driver=none ubuntu dd if=/dev/zero of=/proc/self/fd/1 bs=1M count=1000``` finishes,  Which cause next run to use up memory, (is it because of we rely on golang GC?)

**Comment by unclejack** [comment]
@thefallentree Memory isn't released immediately because of Go's GC.

**Comment by thefallentree** [comment]
@unclejack  which is a problem for us, as we are expecting to launch/destroy many commands which output quite a bit of content to stdout/stderr.  With current implementation, it is likely we need to triple the memory usage (1 in process before writing out, 1 in docker daemon , 1 in client reading) , which is ... a lot.

**Comment by thefallentree** [comment]
@unclejack  on a side note, we are currently executing these commands through SSH,  switching to docker exec make it easier to specify resource constraints on these commands (and better balancing etc), but apparently it is causing all these issues. 

Can you also comment on the PTY thingy?

**Comment by unclejack** [comment]
@thefallentree I'm going to look into this and comment afterwards on that. Please keep in mind I'll have to work on this before being able to provide a fix.

**Comment by thefallentree** [comment]
@unclejack could you take this bug?

**Comment by jonathanperret** [comment]
Chiming in as I've just been hitting this while applying the following pretty useful pattern:

```
docker run --log-driver=none --rm <build-image> bash -xec '
  (git clone some/repo && make)>&2
  tar c build/
' | tar xv
```

i.e., run a tarball-producing command in an ephemeral container, using standard output to route that to a matching `tar` on the client side.

I'm mentioning that in case a reasonable use-case was needed to justify "flooding" stdout.

Seen from my window, blocking the contained process's writes to stdout seems perfectly reasonable and consistent with what would happen if I replaced `docker run … bash` above with a `ssh … bash` equivalent. In fact, I can see only three alternatives for dealing with a reader that reads slower than the container can generate data:

1. buffer it with no limit (current solution, which quickly leads to out-of-memory errors)
2. buffer within a reasonable limit, then block (basically what UNIX pipes do)
3. buffer within a reasonable limit, then drop data

Based on my experience, I would favor capping the daemon's transfer buffer to a reasonable value (a couple megabytes at most, particularly if this is to be multiplied by the number of active clients) and blocking stdout/stderr once that fills up ; and yet @unclejack , in https://github.com/docker/docker/issues/9139#issuecomment-87323320 you mentioned that "Some applications don't handle [getting blocked on stdout or stderr] properly and they crash." I find this surprising (any file descriptor can be expected to block at least some of the time!) but I guess this means Docker would need some kind of flag to support both scenarios: an app that is OK with blocking and wants to spit out lots of data, vs one that absolutely must not be slowed down. Right?

**Comment by unclejack** [comment]
#16190 will help improve this.

**Comment by jonathanperret** [comment]
Thanks @unclejack for the pointer. I'll go ahead and build #16190 to see if it improves my workloads. I'm still a bit puzzled that you would resist capping the buffer, given that the alternative is crashing the daemon (or making it so big that it can't `fork` anymore, which is pretty much the same for practical purposes). But I'll defer to your experience on the matter.

Do you have examples in mind of applications that would behave badly given a blocked stdout/stderr, as you alluded to in https://github.com/docker/docker/issues/9139#issuecomment-87323320? I'd like to help: would testing those with various alternative buffering strategies be useful?

**Comment by unclejack** [comment]
@jonathanperret I'm not opposed to it in general, just opposed to making too many changes at once. The goal is to see if #16190 is 100% OK and take it from there as far as I'm concerned.

**Comment by jonathanperret** [comment]
Understood, sounds like a good plan. Thanks for the clarification.

**Comment by LK4D4** [comment]
Fixed in #17877 
Feel free to try master builds!
